File: frontend/src/types/pcr.ts
export interface MedicationGiven {
  drug: string;
  dose?: number;
  unit?: string;
  route?: string;
  time?: string;
}

export interface PCRDocument {
  age?: number;
  sex?: string;
  chief_complaint?: string;
  primary_impression?: string;
  secondary_impression?: string;
  signs_symptoms: string[];
  events_leading?: string;
  allergies: string[];
  medications_current: string[];
  past_medical_history: string[];
  bp_systolic?: number;
  bp_diastolic?: number;
  heart_rate?: number;
  respiratory_rate?: number;
  spo2?: number;
  gcs_total?: number;
  avpu?: string;
  pain_scale?: number;
  temperature?: number;
  blood_glucose?: number;
  etco2?: number;
  cardiac_rhythm?: string;
  medications_given: MedicationGiven[];
  procedures: string[];
  narrative_text?: string;
}

export interface FieldConfidence {
  value: unknown;
  confidence: number;
  source: 'asr_extraction' | 'user_correction' | 'gap_fill';
  timestamp: string;
  extraction_model: string;
}

export interface PCRStateEnvelope {
  session_id: string;
  pcr: PCRDocument;
  field_confidence: Record<string, FieldConfidence>;
  missing_mandatory: string[];
  missing_required: string[];
  completeness_score: number;
  last_updated: string;
  version: number;
}

File: backend/app/schemas/pcr.py
"""PCR document models with confidence tracking."""

from datetime import datetime, timezone
from typing import Any, Optional

from pydantic import BaseModel, Field


class MedicationGiven(BaseModel):
    """A medication administered during the encounter."""

    drug: str
    dose: Optional[float] = None
    unit: Optional[str] = None
    route: Optional[str] = None
    time: Optional[str] = None


class PCRDocument(BaseModel):
    """Full PCR with all extractable fields aligned to NEMSIS."""

    # Demographics (ePatient)
    age: Optional[int] = None
    sex: Optional[str] = None

    # Incident (eScene/eDispatch)
    incident_location: Optional[str] = None  # eScene.15
    initial_acuity: Optional[str] = None     # eDispatch.13
    protocol_used: Optional[str] = None      # eProtocols.01

    # Situation (eSituation)
    chief_complaint: Optional[str] = None
    primary_impression: Optional[str] = None
    secondary_impression: Optional[str] = None
    signs_symptoms: list[str] = Field(default_factory=list)
    events_leading: Optional[str] = None

    # History (eHistory)
    allergies: list[str] = Field(default_factory=list)
    medications_current: list[str] = Field(default_factory=list)
    past_medical_history: list[str] = Field(default_factory=list)

    # Vitals (eVitals)
    bp_systolic: Optional[int] = None
    bp_diastolic: Optional[int] = None
    heart_rate: Optional[int] = None
    respiratory_rate: Optional[int] = None
    spo2: Optional[int] = None
    gcs_total: Optional[int] = None
    avpu: Optional[str] = None
    pain_scale: Optional[int] = None
    temperature: Optional[float] = None
    blood_glucose: Optional[int] = None
    etco2: Optional[int] = None
    cardiac_rhythm: Optional[str] = None

    # Medications (eMedications)
    medications_given: list[MedicationGiven] = Field(default_factory=list)

    # Procedures (eProcedures)
    procedures: list[str] = Field(default_factory=list)

    # Narrative (eNarrative)
    narrative_text: Optional[str] = None


class FieldConfidence(BaseModel):
    """Confidence metadata for a single PCR field."""

    value: Any
    confidence: float = Field(ge=0.0, le=1.0)
    source: str  # "asr_extraction", "user_correction", "gap_fill"
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    extraction_model: str  # "finetuned_t5", "llm_baseline", "manual"


class PCRStateEnvelope(BaseModel):
    """PCR document wrapped with per-field confidence and completeness metadata."""

    session_id: str
    pcr: PCRDocument
    field_confidence: dict[str, FieldConfidence] = Field(default_factory=dict)
    missing_mandatory: list[str] = Field(default_factory=list)
    missing_required: list[str] = Field(default_factory=list)
    completeness_score: float = 0.0
    last_updated: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    version: int = 0
